compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp

```diff
@@ -1586,5 +1588,55 @@ transform_dialect::CreateMatmulMfmaTileSizesOp::apply(
   return DiagnosedSilenceableFailure::success();
 }
 
+//===----------------------------------------------------------------------===//
+// SetContractionLayoutAttributes
+//===----------------------------------------------------------------------===//
+
+DiagnosedSilenceableFailure
+transform_dialect::SetContractionLayoutAttributes::apply(
+    transform::TransformRewriter &rewriter,
+    transform::TransformResults &results, transform::TransformState &state) {
+  auto payloadList = state.getPayloadOps(getTarget());
+  auto typeList = state.getParams(getMmaType());
+  if (typeList.size() != 1) {
+    return emitDefiniteFailure()
+           << "invalid more than one attribute for contraction annotation";
+  }
+  auto mmaType = llvm::dyn_cast<IREE::GPU::MmaAttr>(typeList.front());
+  if (!mmaType) {
+    return emitDefiniteFailure()
+           << "invalid non-mma attribute for contraction annotation "
+           << typeList.front();
+  }
+
+  for (Operation *payload : payloadList) {
+    auto contract = llvm::dyn_cast<vector::ContractionOp>(payload);
+    if (!contract) {
+      return emitDefiniteFailure()
+             << "invalid non-contraction annotation " << payload;
+    }
+
+    auto maybeLayouts = mmaType.getContractionLayout(contract);
+    if (failed(maybeLayouts)) {
+      return emitDefiniteFailure()
+             << "invalid opaque mma layout for annotation " << mmaType;
+    }
+
+    auto [aLayout, bLayout, cLayout] = *maybeLayouts;
```

Comment:
Note that for nested layout contraction we are also expecting to attach the mfma attribute directly https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPUDistributeContract.cpp#L73.

Reply:
We probably need to refactor this once we switch to the nested layout. Maybe that means shifting the current logic to this transform op.